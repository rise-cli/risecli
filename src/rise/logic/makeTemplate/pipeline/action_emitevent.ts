export function buildEventDetailInput(input: any) {
    let detail = `"{`
    Object.keys(input).forEach((x, i, array) => {
        let value = input[x].toString()

        if (value.startsWith('$')) {
            value = value.replace('$', '$ctx.args.input.')
        }

        if (value.startsWith('!')) {
            value = value.replace('!', '$ctx.identity.claims.')
        }

        if (value.startsWith('#')) {
            value = value.replace('#', '$ctx.stash.dbresult.')
        }

        const EVENT_QUOTE = 'RISE_EVENT_QUOTE'
        detail =
            detail +
            `${EVENT_QUOTE}${x}${EVENT_QUOTE}: ${EVENT_QUOTE}${value}${EVENT_QUOTE}`

        if (i + 1 < array.length) {
            detail = detail + ','
        }
    })
    detail = detail + `}"`
    return detail
}

export function makeEmitEvent(props: any) {
    let detail = buildEventDetailInput(props.config.detail)
    return {
        Resources: {
            [`FunctionEmit${props.name}`]: {
                Type: 'AWS::AppSync::FunctionConfiguration',
                Properties: {
                    ApiId: {
                        'Fn::GetAtt': ['GraphQlApi', 'ApiId']
                    },
                    DataSourceName: 'EventBridgeDataSource',
                    FunctionVersion: '2018-05-29',
                    Name: `FunctionEmit${props.name}`,
                    RequestMappingTemplate: `
                    {
                        "version": "2018-05-29",
                        "method": "POST",
                        "resourcePath": "/",
                        "params": {
                        "headers": {
                            "content-type": "application/x-amz-json-1.1",
                            "x-amz-target":"AWSEvents.PutEvents"
                        },
                        "body": {
                            "Entries":[ 
                                    {
                                        "Source":"custom.${props.config.source}",
                                        "EventBusName": "${props.config.eventBus}",
                                        "Detail": ${detail},
                                        "DetailType":"${props.config.event}"
                                    }
                                ]
                            }
                        }
                    }`,
                    ResponseMappingTemplate: `
                    #if($ctx.error)
                        $util.error($ctx.error.message, $ctx.error.type)
                    #end
   
                    #if($ctx.result.statusCode == 200)
                        $util.qr($ctx.stash.put("eventresult", $ctx.result.body))
                        ## If response is 200, return the body.
                        {
                            "result": "$util.parseJson($ctx.result.body)"
                        }    
                    #else
                        $utils.appendError($ctx.result.body, $ctx.result.statusCode)
                    #end`
                },
                DependsOn: ['EventBridgeDataSource', 'GraphQlSchema']
            }
        }
    }
}
