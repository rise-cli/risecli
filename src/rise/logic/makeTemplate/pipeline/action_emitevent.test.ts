import { buildEventDetailInput, makeEmitEvent } from './action_emitevent'

test('buildEventDetailInput will create a valid event detail object', () => {
    const input = {
        one: '1',
        two: 2
    }

    const result = buildEventDetailInput(input)
    expect(result).toBe('"{\\"one\\": \\"1\\",\\"two\\": \\"2\\"}"')
})

test('buildEventDetailInput will handle special characters correctly', () => {
    const input = {
        one: '!sub',
        two: '$input',
        three: '#dbvalue'
    }

    const result = buildEventDetailInput(input)
    expect(result).toBe(
        '"{\\"one\\": \\"$ctx.identity.claims.sub\\",\\"two\\": \\"$ctx.args.input.input\\",\\"three\\": \\"$ctx.stash.dbresult.dbvalue\\"}"'
    )
})

test('buildEventDetailInput will handle special characters correctly', () => {
    const input = {
        name: 'myevent',
        config: {
            source: 'source',
            eventBus: 'eventBus',
            event: 'event',
            detail: {
                one: '!sub',
                two: '$input',
                three: '#dbvalue'
            }
        }
    }

    const result = makeEmitEvent(input)
    expect(result).toMatchSnapshot()
})
