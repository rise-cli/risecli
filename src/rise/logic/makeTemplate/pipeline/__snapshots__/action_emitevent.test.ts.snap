// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildEventDetailInput will handle special characters correctly 1`] = `
Object {
  "Resources": Object {
    "FunctionEmitmyevent": Object {
      "DependsOn": Array [
        "EventBridgeDataSource",
        "GraphQlSchema",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GraphQlApi",
            "ApiId",
          ],
        },
        "DataSourceName": "EventBridgeDataSource",
        "FunctionVersion": "2018-05-29",
        "Name": "FunctionEmitmyevent",
        "RequestMappingTemplate": "
                    {
                        \\"version\\": \\"2018-05-29\\",
                        \\"method\\": \\"POST\\",
                        \\"resourcePath\\": \\"/\\",
                        \\"params\\": {
                        \\"headers\\": {
                            \\"content-type\\": \\"application/x-amz-json-1.1\\",
                            \\"x-amz-target\\":\\"AWSEvents.PutEvents\\"
                        },
                        \\"body\\": {
                            \\"Entries\\":[ 
                                    {
                                        \\"Source\\":\\"custom.source\\",
                                        \\"EventBusName\\": \\"eventBus\\",
                                        \\"Detail\\": \\"{\\\\\\"one\\\\\\": \\\\\\"$ctx.identity.claims.sub\\\\\\",\\\\\\"two\\\\\\": \\\\\\"$ctx.args.input.input\\\\\\",\\\\\\"three\\\\\\": \\\\\\"$ctx.stash.dbresult.dbvalue\\\\\\"}\\",
                                        \\"DetailType\\":\\"event\\"
                                    }
                                ]
                            }
                        }
                    }",
        "ResponseMappingTemplate": "
                    #if($ctx.error)
                        $util.error($ctx.error.message, $ctx.error.type)
                    #end
   
                    #if($ctx.result.statusCode == 200)
                        $util.qr($ctx.stash.put(\\"eventresult\\", $ctx.result.body))
                        ## If response is 200, return the body.
                        {
                            \\"result\\": \\"$util.parseJson($ctx.result.body)\\"
                        }    
                    #else
                        $utils.appendError($ctx.result.body, $ctx.result.statusCode)
                    #end",
      },
      "Type": "AWS::AppSync::FunctionConfiguration",
    },
  },
}
`;
